hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -221,48 +218,79 @@ public AzureIngressHandler getIngressHandler() {
+   * <p>
+   * If the `ingressHandler` is already initialized and the switch operation is complete, the existing
+   * handler is returned without acquiring a lock to minimize performance overhead.
+   * If the `ingressHandler` is initialized but the switch is incomplete, a lock is acquired to safely modify
+   * or create a new handler. Double-checked locking is used to ensure thread safety while minimizing the
+   * time spent in the critical section.
+   * If the `ingressHandler` is `null`, the handler is safely initialized outside of the lock as no other
+   * thread would be modifying it.
    *
-   * @param serviceType the type of service, either BLOB or DFS.
-   * @param blockFactory the factory to create data blocks.
-   * @param bufferSize the buffer size for the ingress handler.
-   * @return the created {@link AzureIngressHandler}.
-   * @throws IOException if there is an error creating the ingress handler.
+   * @param serviceType   The type of Azure service to handle (e.g., ABFS, Blob, etc.).
+   * @param blockFactory  The factory to create data blocks used in the handler.
+   * @param bufferSize    The buffer size used by the handler for data processing.
+   * @param isSwitch      A flag indicating whether a switch operation is in progress.
+   * @param blockManager  The manager responsible for handling blocks of data during processing.
+   *
+   * @return The initialized or existing Azure ingress handler.
+   * @throws IOException If an I/O error occurs during handler creation or data processing.
    */
   private AzureIngressHandler createIngressHandler(AbfsServiceType serviceType,
       DataBlocks.BlockFactory blockFactory,
       int bufferSize, boolean isSwitch, AzureBlockManager blockManager) throws IOException {
-    lock.lock();
-    try {
-      this.client = clientHandler.getClient(serviceType);
-      if (switchCompleted && ingressHandler != null) {
-        return ingressHandler; // Return the existing ingress handler
-      }
-      if (isDFSToBlobFallbackEnabled
-          && serviceTypeAtInit == AbfsServiceType.BLOB) {
-        throw new InvalidConfigurationValueException(
-            "The ingress service type must be configured as DFS");
-      }
-      if (isDFSToBlobFallbackEnabled && !isSwitch) {
-        ingressHandler = new AzureDfsToBlobIngressFallbackHandler(this,
-            blockFactory,
-            bufferSize, eTag, clientHandler);
-      } else if (serviceType == AbfsServiceType.BLOB) {
-        ingressHandler = new AzureBlobIngressHandler(this, blockFactory,
-            bufferSize, eTag, clientHandler, blockManager);
-      } else {
-        ingressHandler = new AzureDFSIngressHandler(this, blockFactory,
-            bufferSize, eTag, clientHandler);
-      }
-      if (isSwitch) {
-        switchCompleted = true;
-      }
+    // If the ingressHandler is already initialized and switch is complete, no need to acquire a lock
+    if (ingressHandler != null && switchCompleted) {
```

Comment:
These multiple checks are needed to prevent locking, if we use this approach we would be locking it even when switch has already been completed or ingress handler was already initialized

Reply:
taken